### PR TITLE
Bike keys and registration plates

### PR DESCRIPTION
**Pull request by DreamySkrell**
image

- bikes have registration plates
- need the bike's key to start and ride it, can't use any random key, or no key at all
- ~~can steal someone's key while they are riding their bike~~
- ~~can steal a key and throw it in the ocean so they can't ride it ever again~~
- a bike can spawn with a random registration plate and key in ignition
- a bike can spawn without a key, but have its key mapped in somewhere else on the map (or in a rental bike vendor introduced in a different PR)

this does not change mapping or add a way to rent bikes - that will come in another PR

### CODE REVIEW

**Comment by DreamySkrell** [comment]
!review

File: code/modules/vehicles/bike.dm [hunk 1]
```diff
@@ -30,6 +33,21 @@
...
+	/// Registration plate string of the vehicle, visible on examine,
+	/// to distingush different vehicles of the same type from each other.
+	/// Also used to check if the key is for this vehicle.
+	/// If null, it is randomly generated on init.
+	var/registration_plate = null
+	/// Key type accepted in vehicle ignition.
+	var/key_type = /obj/item/key/bike
+	/// Actual key object in the vehicle ignition, or null if no key in ignition.
+	/// To actually start the vehicle, key data needs to match with the registration plate string.
+	var/obj/item/key/key = null
```
Comment (FluffyGhoster) [inline code/modules/vehicles/bike.dm:45]: You need to QDEL it in Destroy

File: code/modules/vehicles/bike.dm [hunk 2]
```diff
@@ -105,16 +145,33 @@
...
+		if(on)
+			toggle_engine(user)
 	else if(user != load && load)
 		user.visible_message ("[user] starts to unbuckle [load] from \the [src]!")
 		if(do_after(user, 8 SECONDS, src))
 			unload(load)
 			to_chat(user, "You unbuckle [load] from \the [src]")
 			to_chat(load, "You were unbuckled from \the [src] by [user]")
 
+/obj/vehicle/bike/attackby(obj/item/item, mob/living/user)
```
Comment (FluffyGhoster) [inline code/modules/vehicles/bike.dm:161]: Preserve the header like every other attackby proc

File: code/modules/vehicles/bike.dm [hunk 3]
```diff
@@ -50,9 +85,15 @@
 		return
 
 	if(!on)
-		turn_on()
-		src.visible_message("\The [src] rumbles to life.", "You hear something rumble deeply.")
-		playsound(src, 'sound/machines/vehicles/bike_start.ogg', 100, 1)
+		if(!key)
+			to_chat(user, SPAN_WARNING("You cannot turn \the [src] on, without a key."))
+		else if(key && (key.key_data != registration_plate))
```
Comment (FluffyGhoster) [inline code/modules/vehicles/bike.dm:90]:
```suggestion
			to_chat(user, SPAN_WARNING("You cannot turn \the [src] on, without a key."))
			return

		if((key.key_data != registration_plate))
```

**Comment by Greenjoe12345** [comment]
This'd want to be disabled on the Climber. Also how does it interact with the coin operated snowmobile from the casino?

File: code/modules/vehicles/bike.dm (hunk 1 above)
Comment (DreamySkrell) [reply code/modules/vehicles/bike.dm:45]: done. QDEL_NULL, right?

File: code/modules/vehicles/bike.dm (hunk 2 above)
Comment (DreamySkrell) [reply code/modules/vehicles/bike.dm:165]: done, if you mean to the item var should be `attacking_item`

Comment (FluffyGhoster) [reply code/modules/vehicles/bike.dm]: not only, the second parameter is a mob/user, not a mob/living/user

Comment (DreamySkrell) [reply code/modules/vehicles/bike.dm]: done

**Review by FluffyGhoster — APPROVED** [review]

**Review by NonQueueingMatt — APPROVED** [review]

**Pull request merged**